### PR TITLE
Update to oneTBB 2021.4.0

### LINKED ISSUE #1424: TBB back-end needs to be migrated to oneTBB
Since TBB is now also under the oneAPI umbrella some interfaces have changed / were deprecated. We should update the back-end to be compatible with both versions until the old style TBB is no longer required.

https://docs.oneapi.io/versions/latest/onetbb/tbb_userguide/Migration_Guide.html

### PR DESCRIPTION
**Pull request by bernhardmgruber**
This PR updates to oneTBB 2021.4.0.

### CODE REVIEW

File: CMakeLists.txt
Comment (bernhardmgruber) [inline CMakeLists.txt]: I don't know if we still need this. I think I saw a reference to HIP somewhere? Does HIP also need a CMake module?

File: script/install_tbb.sh
Comment (j-stephan) [inline script/install_tbb.sh]: The version number does not match the title of the PR.

**Review by j-stephan — CHANGES_REQUESTED** [review]
Since you are touching the TBB installation already: Do you think you can fix #716 along the way?

File: script/install_tbb.sh
Comment (bernhardmgruber) [reply script/install_tbb.sh]: Thx! I had to change it after creating the PR, because cmake could not find 2021.1.1 yet. I think the full cmake support came in 2021.2.0.

File: CMakeLists.txt
Comment (j-stephan) [reply CMakeLists.txt]: ROCm comes with its own `FindHIP.cmake` nowadays and a user should point to its location on the command line. IMO this can be removed, too.

Comment (bernhardmgruber) [reply CMakeLists.txt]: So users will never define an environment variable `CMAKE_MODULE_PATH` to inject `FindHIP.cmake`?

Comment (bernhardmgruber) [reply CMakeLists.txt]: Ok, removed.

Comment (j-stephan) [reply CMakeLists.txt]: To expand a bit on this: The user should provide the correct path by setting one of the following:

1. The `<PackageName>_ROOT` environment variable / CMake variable.
2. The `CMAKE_PREFIX_PATH` CMake variable.
3. The `<PackageName>_DIR` environment variable.
4. The user package registry - [Link](https://cmake.org/cmake/help/v3.18/manual/cmake-packages.7.html#user-package-registry).

The top overrides the bottom, i.e. `<PackageName>_ROOT` will override anything else, `CMAKE_PREFIX_PATH` will override `<PackageName>_DIR` and the package registry, and so on.

Comment (bernhardmgruber) [reply CMakeLists.txt]: thx for the explanation!

**Comment by bernhardmgruber** [comment]
I really don't understand why the variable `ALPAKA_CI_TBB_VERSION` from the `ci.yml` does not arrive in the `install_tbb.sh` script. Does anyone have an idea?

File: .github/workflows/ci.yml
Comment (j-stephan) [inline .github/workflows/ci.yml]: You probably want to set the TBB version here as well, otherwise the CI job won't know about it.

Comment (bernhardmgruber) [reply .github/workflows/ci.yml]: Discussed offline, the job should inherit the environment form the matrix.

**Comment by bernhardmgruber** [comment]
> I really don't understand why the variable `ALPAKA_CI_TBB_VERSION` from the `ci.yml` does not arrive in the `install_tbb.sh` script. Does anyone have an idea?

Found it. We need to specify all variables from the Github actions environment again for the docker image.

**Comment by bernhardmgruber** [comment]
Turns out the Intel apt repository is unresponsive quite often, causing a lot of surious CI failures.

**Comment by j-stephan** [comment]
I think I saw a guide somewhere which tells you how to cache binaries (and other stuff). Maybe we should look at that in the future.

**Comment by bernhardmgruber** [comment]
> I think I saw a guide somewhere which tells you how to cache binaries (and other stuff). Maybe we should look at that in the future.

I already tried that for the CUDA installer once, thinking it would make the CI faster. But actually, restoring the cache took longer than downloading the CUDA installer again. So I am afraid that would slow down the CI even more. It might solve the unresponsiveness of the Intel apt repo though. Hmm.

**Comment by bernhardmgruber** [comment]
It seems we cannot combine nvcc 11.1-11.4, g++-9 and oneTBB since nvcc does not support some gcc builtins that are somehow pulled in via avx512 intrinsics.

**Comment by bernhardmgruber** [comment]
It turns out that the gcc builtin was removed in g++-10 and also libstdc++ from g++-10 is no longer using it. Could it be that nvcc in the failing builds uses a different host compiler than the CXX compiler and libstdc++?

**Comment by bernhardmgruber** [comment]
It seems we somehow get conversion errors now in the OMP atomics:
```
/home/runner/work/alpaka/alpaka/include/alpaka/atomic/AtomicOmpBuiltIn.hpp:42:17: error: conversion from 'int' to 'short int' may change value [-Werror=conversion]
   42 | #        pragma omp atomic capture
      |                 ^~~
```

File: .github/workflows/ci.yml
Comment (bernhardmgruber) [inline .github/workflows/ci.yml:213]: TBB is turned off here, because it is not compatible with libc++ (would need to compile TBB with libc++).

File: .github/workflows/ci.yml
Comment (bernhardmgruber) [inline .github/workflows/ci.yml:130]: Enable TBB here, so we have a non-nvcc windows build.

File: .github/workflows/ci.yml
Comment (bernhardmgruber) [inline .github/workflows/ci.yml:102]: libc++ not compatible with TBB.

File: include/alpaka/atomic/AtomicOmpBuiltIn.hpp
Comment (bernhardmgruber) [inline include/alpaka/atomic/AtomicOmpBuiltIn.hpp]: For some reason, including TBB suddenly generated warnings here, which were turned into errors.

File: script/travis_retry.sh
Comment (bernhardmgruber) [inline script/travis_retry.sh:43]: I was surprised this actually worked. The Intel apt repository is sometimes not available and `sudo apt add-repository` fails after 3 attempts. But trying more often actually works and does not noticably slow down the CI. I am open to have a different `max` here, but `3` is definitely too little :)

File: .github/workflows/ci.yml
Comment (bernhardmgruber) [inline .github/workflows/ci.yml:349]: All other nvcc/g++ builds are release, so I changed one to debug to also test that. I think I wanted to try something because of the failing AVX512 intrinsics in release mode. It is unrelated to adding TBB, but I would keep it.

File: include/alpaka/atomic/AtomicOmpBuiltIn.hpp
Comment (psychocoderHPC) [reply include/alpaka/atomic/AtomicOmpBuiltIn.hpp]: I am not sure if a cast within the capture is allowed. It is not defined by the standard. IMO only a binary operator is allowed.
https://www.openmp.org/spec-html/5.0/openmpsu95.html

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]
I can not merge this change because I do not know if this and other changes of the OpenMP atomics are standard conform.

File: include/alpaka/atomic/AtomicOmpBuiltIn.hpp
Comment (psychocoderHPC) [reply include/alpaka/atomic/AtomicOmpBuiltIn.hpp]: IMO only

```
{ v = x; x binop= expr; } 
{ x binop= expr; v = x; } 
{ v = x; x = x binop expr; } 
{ v = x; x = expr binop x; } 
{ x = x binop expr; v = x; } 
{ x = expr binop x; v = x; } 
{ v = x; x = expr; } 
{ v = x; x++; } 
{ v = x; ++x; } 
{ ++x; v = x; } 
{ x++; v = x; } 
{ v = x; x--; } 
{ v = x; --x; } 
{ --x; v = x; } 
{ x--; v = x; }
```

Is allowed.

File: script/travis_retry.sh
Comment (psychocoderHPC) [reply script/travis_retry.sh:43]: An ideal magic number would be `42` ^^

**Comment by bernhardmgruber** [comment]
> I can not merge this change because I do not know if this and other changes of the OpenMP atomics are standard conform.

I see. Then I have to disable the warnings locally.

File: include/alpaka/atomic/AtomicOmpBuiltIn.hpp
Comment (bernhardmgruber) [reply include/alpaka/atomic/AtomicOmpBuiltIn.hpp]: Replaced by pragmas to disable the warnings.

**Review by j-stephan — APPROVED** [review]

**Comment by j-stephan** [comment]
Merging is blocked until @psychocoderHPC gives his approval.

**Review by psychocoderHPC — APPROVED** [review]

**Pull request merged**
